subsys/shell/shell.c

```diff
@@ -1321,10 +1321,12 @@ void shell_thread(void *shell_handle, void *arg_log_backend,
 					   log_level);
 	}
 
-	/* Enable shell and print prompt. */
-	err = shell_start(shell);
-	if (err != 0) {
-		return;
+	if (IS_ENABLED(CONFIG_SHELL_AUTOSTART)) {
```

Comment:
i wonder if this shouldn't be done on shell instance level (e.g. shell uart option). It's not obligatory to start shell instance at boot, it was specific shell instance implementation choice (e.g. uart shell). There are[ out-of-tree instances](https://github.com/nrfconnect/sdk-nrf/blob/7bb844232d251eb2922b049b8f5d576ad1c557d5/subsys/shell/shell_bt_nus.c#L165) which do not start at boot (it defers shell thread creation) and disabling this option will break them.

Imo, better would be to add an option to uart shell (if it is about uart) to not use `SYS_INIT` but expose `enable_shell_uart()` function in the `shell_uart.h` api (rename it to `shell_uart_enable`).

Reply:
Right now shell has an interface `shell_start` and `shell_stop`, so this change is not breaking anything because the application can call those functions, also by default autostart is enabled.

 But I think that your proposal is better matching to my needs. I will prepare a pull request for it.

Reply:
@nordic-krch based on your proposal I created https://github.com/zephyrproject-rtos/zephyr/pull/51551

Reply:
@nordic-krch I think this solution is more generic than not using `SYS_INIT` in particular backend. First of all AUTOSTART is enabled by default, so it should not break anything as is now. And when you consciously disable it you must always remember to manually enable shell whether init is deferred or not.
In this solution all resources including thread are initialised correctly where they are now, but shell thread waits on `k_poll` which should be safe. Then you can enable backends as you wish. Do not need to init them manually, because you can safely assume they are all correctly initialised, as name suggest enable not init. Not initialising backend in sys_init is far less intuitive and generic, I think.